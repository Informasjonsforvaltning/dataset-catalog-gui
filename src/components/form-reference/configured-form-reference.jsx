import { reduxForm } from 'redux-form';

import FormReference from './form-reference.component';
import validate from './form-reference-validations';
import { asyncValidateDatasetInvokePatch } from '../../pages/dataset-registration-page/formsLib/asyncValidateDatasetInvokePatch';

const config = {
  form: 'reference',
  validate,
  asyncValidate: asyncValidateDatasetInvokePatch
};

export const ConfiguredFormTitle = reduxForm(config)(FormReference);
