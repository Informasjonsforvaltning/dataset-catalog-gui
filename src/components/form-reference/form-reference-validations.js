import * as yup from 'yup';

import TranslationsService from '../../services/translations';

export default yup.object().shape({
  relations: yup.array().of(
    yup.object().shape({
      uri: yup
        .string()
        .url(TranslationsService.translate('validation.validateLink'))
    })
  )
});
