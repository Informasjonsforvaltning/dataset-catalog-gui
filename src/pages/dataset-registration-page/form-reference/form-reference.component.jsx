import React from 'react';
import PropTypes from 'prop-types';
import { Field, FieldArray } from 'redux-form';
import localization from '../../../services/localization';
import { Helptext } from '../../../components/helptext/helptext.component';
import SelectField from '../../../components/fields/field-select/field-select.component';
import { datasetFormPatchThunk } from '../formsLib/asyncValidateDatasetInvokePatch';
import { getTranslateText } from '../../../services/translateText';

const renderReadOnly = ({ input, referenceDatasetsItems }) => {
  const { referenceType } = input.value || {};
  const referenceTypeText = getTranslateText(
    referenceType && referenceType.prefLabel
  );

  const uri = input.value && input.value.source && input.value.source.uri;

  const dataset = referenceDatasetsItems.find(i => i.uri === uri);

  const datasetTitle = getTranslateText(dataset && dataset.title);

  return (
    <div className="pl-3">
      {referenceTypeText} {datasetTitle}
    </div>
  );
};
renderReadOnly.propTypes = {
  input: PropTypes.object.isRequired,
  referenceDatasetsItems: PropTypes.array.isRequired
};

const renderReferenceFields = ({
  item,
  index,
  fields,
  referenceTypesItems,
  referenceDatasetsItems,
  onDeleteFieldAtIndex,
  isReadOnly
}) => (
  <div key={item}>
    {isReadOnly && (
      <Field
        name={`${item}`}
        component={renderReadOnly}
        referenceDatasetsItems={referenceDatasetsItems}
      />
    )}
    {!isReadOnly && (
      <div className="d-flex mb-2" key={index}>
        <div className="w-50">
          <Field
            name={`${item}.referenceType`}
            component={SelectField}
            items={referenceTypesItems}
          />
        </div>
        <div className="w-50">
          <Field
            name={`${item}.source`}
            component={SelectField}
            items={referenceDatasetsItems}
            labelKey="title"
          />
        </div>
        <div className="d-flex align-items-end">
          <button
            className="fdk-btn-no-border"
            type="button"
            title="Remove reference"
            onClick={() => onDeleteFieldAtIndex(fields, index)}
          >
            <i className="fa fa-trash mr-2" />
          </button>
        </div>
      </div>
    )}
  </div>
);
renderReferenceFields.propTypes = {
  item: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  fields: PropTypes.object.isRequired,
  referenceTypesItems: PropTypes.array.isRequired,
  referenceDatasetsItems: PropTypes.array.isRequired,
  onDeleteFieldAtIndex: PropTypes.func.isRequired,
  isReadOnly: PropTypes.bool.isRequired
};

export const renderReference = ({
  fields,
  referenceTypesItems,
  referenceDatasetsItems,
  onDeleteFieldAtIndex,
  isReadOnly
}) => {
  return (
    <div>
      {fields &&
        fields.map((item, index) =>
          renderReferenceFields({
            item,
            index,
            fields,
            referenceTypesItems,
            referenceDatasetsItems,
            onDeleteFieldAtIndex,
            isReadOnly
          })
        )}
      {!isReadOnly && (
        <button
          className="fdk-btn-no-border"
          type="button"
          onClick={() => fields.push({})}
        >
          <i className="fa fa-plus mr-2" />
          {localization.schema.reference.addReferenceLabel}
        </button>
      )}
    </div>
  );
};
renderReference.propTypes = {
  fields: PropTypes.object.isRequired,
  referenceTypesItems: PropTypes.array.isRequired,
  referenceDatasetsItems: PropTypes.array.isRequired,
  onDeleteFieldAtIndex: PropTypes.func.isRequired,
  isReadOnly: PropTypes.bool.isRequired
};
export const FormReference = props => {
  const { initialValues, dispatch, catalogId, datasetId, isReadOnly } = props;
  const { referenceTypesItems, referenceDatasetsItems } = initialValues;
  const deleteFieldAtIndex = (fields, index) => {
    const values = fields.getAll();
    // use splice instead of skip, for changing the bound value
    values.splice(index, 1);
    const patch = { [fields.name]: values };
    const thunk = datasetFormPatchThunk({ catalogId, datasetId, patch });
    dispatch(thunk);
  };

  if (initialValues) {
    return (
      <form>
        <div className="form-group">
          <Helptext
            title={localization.schema.reference.helptext.reference}
            terms="Dataset_relation"
          />
          <FieldArray
            name="references"
            component={renderReference}
            referenceTypesItems={referenceTypesItems}
            referenceDatasetsItems={referenceDatasetsItems}
            onDeleteFieldAtIndex={deleteFieldAtIndex}
            isReadOnly={isReadOnly}
          />
        </div>
      </form>
    );
  }
  return null;
};

FormReference.defaultProps = {
  initialValues: null,
  dispatch: null,
  catalogId: null,
  datasetId: null,
  isReadOnly: false
};

FormReference.propTypes = {
  initialValues: PropTypes.object,
  dispatch: PropTypes.func,
  catalogId: PropTypes.string,
  datasetId: PropTypes.string,
  isReadOnly: PropTypes.bool
};
