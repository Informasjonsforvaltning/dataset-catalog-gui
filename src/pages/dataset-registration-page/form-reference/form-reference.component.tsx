import React, { memo, FC } from 'react';
import { compose } from 'redux';
import { Field, FieldArray } from 'redux-form';

import {
  withTranslations,
  Props as TranslationsProps
} from '../../../providers/translations';

import Translation from '../../../components/translation';
import Helptext from '../../../components/helptext/helptext.component';
import SelectField from '../../../components/fields/field-select/field-select.component';
import MultilingualField from '../../../components/multilingual-field/multilingual-field.component';
import InputField from '../../../components/fields/field-input/field-input.component';
import InputFieldReadonly from '../../../components/fields/field-input-readonly/field-input-readonly.component';

import { datasetFormPatchThunk } from '../formsLib/asyncValidateDatasetInvokePatch';

interface ExternalProps {
  initialValues: any;
  dispatch: (arg: any) => void;
  catalogId: string;
  datasetId: string;
  isReadOnly: boolean;
  onInputChange: () => void;
  languages: any[];
}

interface Props extends ExternalProps, TranslationsProps {}

const FormReference: FC<Props> = ({
  initialValues: { referenceTypesItems, referenceDatasetsItems },
  dispatch,
  catalogId,
  datasetId,
  languages,
  isReadOnly,
  onInputChange,
  translationsService
}) => {
  const deleteFieldAtIndex = (fields: any, index: number) => {
    const values = fields.getAll();
    // use splice instead of skip, for changing the bound value
    values.splice(index, 1);
    const patch = { [fields.name]: values };
    const thunk = datasetFormPatchThunk({ catalogId, datasetId, patch });
    dispatch(thunk);
  };

  const renderReadOnly = ({ input }: any) => {
    const { referenceType } = input.value || {};
    const referenceTypeText = translationsService.translate(
      referenceType && referenceType.prefLabel
    );

    const uri = input.value && input.value.source && input.value.source.uri;

    const dataset = referenceDatasetsItems.find((i: any) => i.uri === uri);

    const datasetTitle = translationsService.translate(dataset?.title);

    return (
      <div className='pl-3'>
        {referenceTypeText} {datasetTitle}
      </div>
    );
  };

  const renderReferenceFields = ({
    item,
    index,
    fields,
    onDeleteFieldAtIndex
  }: any) => (
    <div key={item}>
      {isReadOnly && (
        <Field
          name={`${item}`}
          component={renderReadOnly}
          referenceDatasetsItems={referenceDatasetsItems}
        />
      )}
      {!isReadOnly && (
        <div className='d-flex mb-2' key={index}>
          <div className='w-50'>
            <Field
              name={`${item}.referenceType`}
              component={SelectField}
              items={referenceTypesItems}
            />
          </div>
          <div className='w-50'>
            <Field
              name={`${item}.source`}
              component={SelectField}
              items={referenceDatasetsItems}
              labelKey='title'
              onInputChange={onInputChange}
            />
          </div>
          <div className='d-flex align-items-end'>
            <button
              className='fdk-btn-no-border'
              type='button'
              title='Remove reference'
              onClick={() => onDeleteFieldAtIndex(fields, index)}
            >
              <i className='fa fa-trash mr-2' />
            </button>
          </div>
        </div>
      )}
    </div>
  );

  const renderRelationFields = ({ field, onDeleteFieldAtIndex }: any) => (
    <div key={field}>
      {isReadOnly && (
        <div className='mb-4'>
          <MultilingualField
            name={`${field}.prefLabel`}
            component={InputFieldReadonly}
            languages={languages}
            label='Title'
            showLabel
          />
          <div className='mt-4'>
            <Field
              name={`${field}.uri`}
              component={InputFieldReadonly}
              label='Lenke'
              showLabel
            />
          </div>
        </div>
      )}
      {!isReadOnly && (
        <div className='mb-4' key={field}>
          <MultilingualField
            name={`${field}.prefLabel`}
            component={InputField}
            languages={languages}
            label='Title'
            showLabel
          />
          <div className='mt-4'>
            <Field
              name={`${field}.uri`}
              component={InputField}
              label='Lenke'
              showLabel
            />
          </div>
          <div className='d-flex align-items-end my-2 mb-4'>
            <button
              className='fdk-btn-no-border'
              type='button'
              title='Remove reference'
              onClick={onDeleteFieldAtIndex}
            >
              <i className='fa fa-trash mr-2' />
              <Translation id='schema.reference.removeRelationLabel' />
            </button>
          </div>
        </div>
      )}
    </div>
  );

  const renderReference = ({ fields, onDeleteFieldAtIndex }: any) => (
    <div>
      {fields?.map((item: any, index: number) =>
        renderReferenceFields({
          item,
          index,
          fields,
          referenceTypesItems,
          referenceDatasetsItems,
          onDeleteFieldAtIndex,
          isReadOnly,
          onInputChange
        })
      )}
      {!isReadOnly && (
        <button
          className='fdk-btn-no-border'
          type='button'
          onClick={() => fields.push({})}
        >
          <i className='fa fa-plus mr-2' />
          <Translation id='schema.reference.addReferenceLabel' />
        </button>
      )}
    </div>
  );

  const renderRelations = ({ fields, onDeleteFieldAtIndex }: any) => (
    <div>
      {fields.map((field: any, index: number) =>
        renderRelationFields({
          field,
          languages,
          onDeleteFieldAtIndex: () => onDeleteFieldAtIndex(fields, index),
          isReadOnly
        })
      )}
      {!isReadOnly && (
        <button
          className='fdk-btn-no-border'
          type='button'
          onClick={() => fields.push({})}
        >
          <i className='fa fa-plus mr-2' />
          <Translation id='schema.reference.addRelationLabel' />
        </button>
      )}
    </div>
  );

  return referenceTypesItems && referenceDatasetsItems ? (
    <form>
      <div className='form-group'>
        <Helptext
          title={translationsService.translate(
            'schema.reference.helptext.reference'
          )}
          term='Dataset_relation'
          recommended
        />
        <FieldArray
          name='references'
          component={renderReference}
          referenceTypesItems={referenceTypesItems}
          referenceDatasetsItems={referenceDatasetsItems}
          onDeleteFieldAtIndex={deleteFieldAtIndex}
          isReadOnly={isReadOnly}
          onInputChange={onInputChange}
        />
      </div>
      <div className='form-group'>
        <Helptext
          title={translationsService.translate(
            'schema.reference.helptext.relatedResources'
          )}
          term='Dataset_relation_resource'
          recommended
        />
        <FieldArray
          name='relations'
          component={renderRelations}
          languages={languages}
          onDeleteFieldAtIndex={deleteFieldAtIndex}
          isReadOnly={isReadOnly}
        />
      </div>
    </form>
  ) : null;
};

export default compose<FC<ExternalProps>>(
  memo,
  withTranslations
)(FormReference);
