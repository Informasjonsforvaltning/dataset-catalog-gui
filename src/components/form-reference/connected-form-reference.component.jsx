import { getFormSyncErrors } from 'redux-form';
import { connect } from 'react-redux';

import { ConfiguredFormReference } from './configured-form-reference';
import { languageType } from '../../schemaTypes';

const mapStateToProps = (
  state,
  {
    datasetItem: {
      references = [languageType],
      relations = [{ uri: '', prefLabel: {} }]
    },
    referenceTypesItems,
    referenceDatasetsItems
  }
) => ({
  initialValues: {
    references,
    referenceTypesItems,
    referenceDatasetsItems,
    relations
  },
  syncErrors: getFormSyncErrors('reference')(state)
});

export const ConnectedFormReference = connect(mapStateToProps)(
  ConfiguredFormReference
);
