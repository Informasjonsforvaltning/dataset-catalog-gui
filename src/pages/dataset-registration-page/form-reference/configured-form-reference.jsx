import { reduxForm } from 'redux-form';

import { FormReference } from './form-reference.component';
import validate from './form-reference-validations';
import { asyncValidateDatasetInvokePatch } from '../formsLib/asyncValidateDatasetInvokePatch';

const config = {
  form: 'reference',
  validate,
  asyncValidate: asyncValidateDatasetInvokePatch
};

export const ConfiguredFormTitle = reduxForm(config)(FormReference);
