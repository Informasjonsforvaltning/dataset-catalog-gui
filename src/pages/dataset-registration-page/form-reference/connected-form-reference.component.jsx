import { getFormSyncErrors } from 'redux-form';
import { connect } from 'react-redux';
import _ from 'lodash';

import { ConfiguredFormTitle } from './configured-form-reference';
import { languageType } from '../../../schemaTypes';

const mapStateToProps = (state, ownProps) => {
  const { datasetItem, referenceTypesItems, referenceDatasetsItems } = ownProps;
  return {
    initialValues: {
      references:
        _.get(datasetItem, 'references', []).length > 0
          ? _.get(datasetItem, 'references')
          : [languageType],
      referenceTypesItems,
      referenceDatasetsItems
    },
    syncErrors: getFormSyncErrors('reference')(state)
  };
};

export const ConnectedFormReference = connect(mapStateToProps)(
  ConfiguredFormTitle
);
