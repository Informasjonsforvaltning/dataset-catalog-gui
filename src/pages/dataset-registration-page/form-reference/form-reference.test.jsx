import React from 'react';
import { shallow } from 'enzyme';
import { FormReference, renderReference } from './form-reference.component';

let defaultProps;
let wrapper;

beforeEach(() => {
  defaultProps = {
    initialValues: {}
  };
  wrapper = shallow(<FormReference {...defaultProps} />);
});

test('should render FormReference correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render renderReference correctly', () => {
  wrapper = shallow(renderReference(defaultProps));
  expect(wrapper).toMatchSnapshot();
});
