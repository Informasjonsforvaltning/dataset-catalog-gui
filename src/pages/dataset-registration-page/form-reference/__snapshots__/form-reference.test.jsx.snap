// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render FormReference correctly 1`] = `
<form>
  <div
    className="form-group"
  >
    <withStateHandlers(Component)
      terms="Dataset_relation"
      title="Relasjoner"
    />
    <FieldArray
      component={[Function]}
      isReadOnly={false}
      name="references"
      onDeleteFieldAtIndex={[Function]}
    />
  </div>
</form>
`;

exports[`should render renderReference correctly 1`] = `
<div>
  <button
    className="fdk-btn-no-border"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fa fa-plus mr-2"
    />
    Legg til relaterte datasett
  </button>
</div>
`;
